### PR TITLE
[Auth] Fix persistence issue when manually setting/using session storage

### PR DESCRIPTION
**Pull request by sam-gc**
Fixes https://github.com/firebase/firebase-js-sdk/issues/5171

`browserSession` persistence is now added to the escalation path of `getAuth()` to match v8. We also now only "upgrade" the persistence if the persistence deems itself "upgradeable" (which only applies to `indexedDB` and `localStorage`)

### CODE REVIEW

File: packages-exp/auth-exp/test/integration/webdriver/persistence.test.ts
Comment (lisajian) [inline packages-exp/auth-exp/test/integration/webdriver/persistence.test.ts:185]: Clarifying q (trying to understand more than anything!): Not catching why this test would be impacted by the fix to session persistence? This test concerns localStorage and indexDB only, right?

**Review by lisajian — COMMENTED** [review]
LGTM - will defer to @yuchenshi / @avolkovi who have more context on the change! Thanks for the fix!

File: packages-exp/auth-exp/test/integration/webdriver/persistence.test.ts
Comment (sam-gc) [reply packages-exp/auth-exp/test/integration/webdriver/persistence.test.ts:185]: This change also makes it so we only look at "available" persistence layers before initializing the SDK. If `indexedDB` is readonly, it would be filtered out early on, so this behavior (migrating from readonly indexedDB to localStorage) doesn't happen anymore.

Comment (lisajian) [reply packages-exp/auth-exp/test/integration/webdriver/persistence.test.ts:185]: Ohh got it, thanks for the clarification!

File: packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts
Comment (yuchenshi) [inline packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts:162]: What if this is undefined?

File: packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts
Comment (yuchenshi) [inline packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts:117]: nit: rename to `availablePresistences`

File: packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts
Comment (yuchenshi) [inline packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts:137]: Please revert this -- the if-statement is intentional to avoid triggering dubious change events with the migration `._set()` below

File: packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts
Comment (yuchenshi) [inline packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts:145]: I'd suggest not assigning to `chosenPersistence` here since it is not the chosen persistence (but rather the persistence to migrate from). It is very confusing that `chosenPersistence` is then overwritten in L162 below. I'd suggest inlining `_shouldAllowMigration` check from below. Or just use a different variable name if you must.

**Review by yuchenshi — APPROVED** [review]
A few comments inline. I totally understand you're trying to improve the readability here -- and in that spirit I think we should split it into different helper functions.

Step 1: Choosing the preferred persistence (the first one that is available)
Step 2: Attempting to locate the user from all persistences (return user and source persistence)
Step 3 (can stay in this function): Migration logic (a. if user is in a persistence that doesn't support migration, use that instead and skip migration, or b. migrate if different)

With that being said, I don't feel like I have the authority for TypeScript readability to block this PR any longer. Approved with comments inline -- feel free to file a bug for later cleanup.

File: packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts
Comment (sam-gc) [reply packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts]: Fixed by just returning the manager if it is

File: packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts
Comment (sam-gc) [reply packages-exp/auth-exp/src/core/persistence/persistence_user_manager.ts]: Renamed to "selectedPersistence" as that accurately reflects what this code is doing

**Comment by sam-gc** [comment]
After some tinkering I think refactoring into helper functions would make this code even more complex.

**Review by yuchenshi — APPROVED** [review]

**Pull request merged**
